solutions/gold/izho-17-LongestBeautifulSequence.mdx

```diff
@@ -0,0 +1,347 @@
+___
+**Example:** $x$ = 01010, $a_i$ = 01101, $b$ = 3
+```
+01010 & 01101 = 01000 -> 1
+
+010   & 011   = 010   -> 1
+
+   10 &    01 =    00 -> 0
+```
+$1 + 0 = \boxed1$.
+___
+Since our DP state fixes $\texttt{l}(x, b)$, we should rewrite this equation to solve for $\texttt{r}(x, b)$: $\texttt{bc}[\texttt{r}(x, b)][\texttt{r}(a_i, b)] = k_i - \texttt{bc}[\texttt{l}(x, b)][\texttt{l}(a_i, b)] \Rightarrow \texttt{r}(x, b) \in \texttt{trans}_{k_i - \texttt{bc}[\texttt{l}(x, b)][\texttt{l}(a_i, b)]}(\texttt{r}(a_i, b))$. Note this restriction matches our $\texttt{dp3}$ restriction of $\texttt{r}(x, b) \in \texttt{trans}_k(j)$, so we have $j = \texttt{r}(a_i, b)$ and $k = k_i - \texttt{bc}[\texttt{l}(x, b)][\texttt{l}(a_i, b)] = k_i - \texttt{bc}[i][\texttt{l}(a_i, b)]$. As for $\texttt{l}(x, b)$, we simply enumerate all $2^b$ possibilities; therefore, $i \in [0, 2^b)$.
+
+**In summary, the transitions to $a_i$ are all states $\texttt{dp3}[x][\texttt{r}(a_i, b)][k_i - \texttt{bc}[x][\texttt{l}(a_i, b)]]$ where $x \in [0, 2^b)$**.
+
+Because the only variable in the state is $x$, the complexity of transition is the number of options for $x$ which is $\boxed{2^b}$.
+
+#### Affected States:
+We now have the length of the longest LBS that ends at $a_i$. What $\texttt{dp3}[x][y][k]$ states does this affect?
+
+Starting with the obvious, $x = \texttt{l}(a_i, b)$. We also have the restriction that $\texttt{r}(a_i, b) \in \texttt{trans}_k(y) \Rightarrow k = \texttt{bc}[\texttt{r}(a_i, b)][y]$. Because $\texttt{r}(a_i, b)$ is a constant, $k$ depends solely on $y$.
+
+**In summary, the states affected by $a_i$ are all states $\texttt{dp}3[\texttt{l}(a_i, b)][y][\texttt{bc}[\texttt{r}(a_i, b)][y]]$**.
+
+$y$ has to have the same amount of bits as $\texttt{r}(a_i, b)$ which has $\log_2{M} - b$ bits. Therefore, $y$ has $2^{\log_2M - b} = \boxed{\frac{M}{2^b}}$.
+
+To sum it all up, the complexity of transition is $\mathcal{O}(2^b)$ and the complexity of affected states is $\mathcal{O}(\frac{M}{2^b})$. To minimize the sum of these complexities, we should equate them. Because $M = 2 ^{20}$, we find the optimal value of $b$ to be $\boxed{10}$.
+
+### Implementation
+In the implementation below:
+- $\texttt{l}(x)$ denotes the 10 leftmost bits of $x$
+- $\texttt{r}(x)$ denotes the 10 righmost bits of $x$
+- $\texttt{dp}[i][j][k]$: what was previously referred to as $\texttt{dp3}$
+
+**Time Complexity:** $\mathcal{O}(N\sqrt{M})$
+
+<Spoiler title="Code">
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using pii = pair<int, int>;
+const int N = 1e5;
+const int B = 10;
+int bc[1 << B][1 << B];  // bc[i][j] = bit_count(i & j)
+struct state {
+	int length = 0, end = 0;
+	state() {}
+	state(int l, int e) : length(l), end(e) {}
+};
```

Comment:
```suggestion
struct state { int length, end; }
```
also rename to `State`

Reply:
wait does this automatically create a constructor? 😅 never knew that

Reply:
might be g++ specific, but most of the structs here take advantage of it, so yeah